File: erisbot/src/main/java/studio/styx/erisbot/utils/ComponentBuilder.java
package studio.styx.erisbot.utils;
import net.dv8tion.jda.api.components.actionrow.ActionRow;
import net.dv8tion.jda.api.components.buttons.Button;
import net.dv8tion.jda.api.components.container.Container;
import net.dv8tion.jda.api.components.container.ContainerChildComponent;
import net.dv8tion.jda.api.components.section.Section;
import net.dv8tion.jda.api.components.separator.Separator;
import net.dv8tion.jda.api.components.textdisplay.TextDisplay;
import net.dv8tion.jda.api.components.thumbnail.Thumbnail;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.interactions.InteractionHook;
import net.dv8tion.jda.api.interactions.commands.SlashCommandInteraction;
import org.jetbrains.annotations.NotNull;

import java.awt.Color;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

public class ComponentBuilder {

    public static class ContainerBuilder {
        private final List<ContainerChildComponent> components = new ArrayList<>();
        private Color accentColor;
        private boolean isEphemeral = false;
        private boolean disallowedMentions = false;

        private ContainerBuilder() {
            // Construtor privado, use métodos estáticos para iniciar
        }

        // Inicia um novo builder
        public static ContainerBuilder create() {
            return new ContainerBuilder();
        }

        // Adiciona um componente genérico (TextDisplay, Separator, Button, etc.)
        public ContainerBuilder add(ContainerChildComponent component) {
            if (component != null) {
                components.add(component);
            }
            return this;
        }

        // Adiciona um texto simples (converte String para TextDisplay automaticamente)
        public ContainerBuilder addText(String text) {
            if (text != null && !text.isEmpty()) {
                components.add(TextDisplay.of(text));
            }
            return this;
        }

        // Adiciona um divisor com espaçamento específico
        public ContainerBuilder addDivider(@NotNull Boolean isLarge) {
            if (isLarge) {
                components.add(Separator.createDivider(Separator.Spacing.LARGE));
            } else {
                components.add(Separator.createDivider(Separator.Spacing.SMALL));
            }
            return this;
        }

        // Adiciona um divisor com espaçamento invisivel
        public ContainerBuilder addDivider(@NotNull Boolean isInvisble, Boolean isLarge) {
            if (isInvisble) {
                if (isLarge) {
                    components.add(Separator.createInvisible(Separator.Spacing.LARGE));
                } else {
                    components.add(Separator.createInvisible(Separator.Spacing.SMALL));
                }
            } else {
                if (isLarge) {
                    components.add(Separator.createDivider(Separator.Spacing.LARGE));
                } else {
                    components.add(Separator.createDivider(Separator.Spacing.SMALL));
                }
            }
            return this;
        }

        public ContainerBuilder addSection(@NotNull String thumbnailUrl, String text) {
            components.add(
                    Section.of(
                            Thumbnail.fromUrl(thumbnailUrl),
                            TextDisplay.of(text)
                    )
            );

            return this;
        }

        public ContainerBuilder addSection(@NotNull Button button, String text) {
            components.add(
                    Section.of(
                            button,
                            TextDisplay.of(text)
                    )
            );
            return this;
        }

        public ContainerBuilder addRow(@NotNull ActionRow row) {
            components.add(row);
            return this;
        }

        // Define a cor de destaque
        public ContainerBuilder withColor(String hexColor) {
            if (hexColor != null && !hexColor.isEmpty()) {
                // Adiciona o # se não tiver
                String colorString = hexColor.startsWith("#") ? hexColor : "#" + hexColor;
                this.accentColor = Color.decode(colorString);
            }
            return this;
        }

        // Define se a mensagem é efêmera
        public ContainerBuilder setEphemeral(boolean ephemeral) {
            this.isEphemeral = ephemeral;
            return this;
        }

        // Constrói o container
        public Container build() {
            Container container = Container.of(this.components);
            if (accentColor != null) {
                container = container.withAccentColor(accentColor);
            }
            return container;
        }

        public ContainerBuilder disableMentions() {
            this.disallowedMentions = true;
            return this;
        }

        // Envia a resposta diretamente
        public void reply(SlashCommandInteraction event) {
            if (this.disallowedMentions) {
                event.replyComponents(build())
                        .useComponentsV2()
                        .setEphemeral(isEphemeral)
                        .setAllowedMentions(EnumSet.noneOf(Message.MentionType.class))
                        .queue();
            } else {
                event.replyComponents(build())
                        .useComponentsV2()
                        .setEphemeral(isEphemeral)
                        .queue();
            }
        }
        public void reply(ButtonInteractionEvent event) {
            if (this.disallowedMentions) {
                event.replyComponents(build())
                        .useComponentsV2()
                        .setEphemeral(isEphemeral)
                        .setAllowedMentions(EnumSet.noneOf(Message.MentionType.class))
                        .queue();
            } else {
                event.replyComponents(build())
                        .useComponentsV2()
                        .setEphemeral(isEphemeral)
                        .queue();
            }
        }

        public void reply(InteractionHook event) {
            if (this.disallowedMentions) {
                event.editOriginalComponents(build())
                        .useComponentsV2()
                        .setAllowedMentions(EnumSet.noneOf(Message.MentionType.class))
                        .queue();
            } else {
                event.editOriginalComponents(build())
                        .useComponentsV2()
                        .queue();
            }
        }

        public void editOriginal(ButtonInteractionEvent event) {
            if (this.disallowedMentions) {
                event.editComponents(build())
                        .useComponentsV2()
                        .setAllowedMentions(EnumSet.noneOf(Message.MentionType.class))
                        .queue();
            } else {
                event.editComponents(build())
                        .useComponentsV2()
                        .setAllowedMentions(EnumSet.noneOf(Message.MentionType.class))
                        .queue();
            }
        }
    }
}

File: common-discord/src/main/kotlin/utils/ComponentBuilder.kt
package utils

import net.dv8tion.jda.api.components.actionrow.ActionRow
import net.dv8tion.jda.api.components.buttons.Button
import net.dv8tion.jda.api.components.container.Container
import net.dv8tion.jda.api.components.container.ContainerChildComponent
import net.dv8tion.jda.api.components.mediagallery.MediaGallery
import net.dv8tion.jda.api.components.mediagallery.MediaGalleryItem
import net.dv8tion.jda.api.components.section.Section
import net.dv8tion.jda.api.components.separator.Separator
import net.dv8tion.jda.api.components.textdisplay.TextDisplay
import net.dv8tion.jda.api.components.thumbnail.Thumbnail
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.EntitySelectInteractionEvent
import net.dv8tion.jda.api.interactions.InteractionHook
import java.awt.Color

object ComponentBuilder {

    // Classe interna com companion object para criar instâncias
    class ContainerBuilder private constructor() {
        private val components = mutableListOf<ContainerChildComponent>()
        private var accentColor: Color? = null
        private var isEphemeral = false
        private var disallowedMentions = false

        // Métodos fluentes
        fun add(component: ContainerChildComponent?) = apply {
            component?.let { components.add(it) }
        }

        fun addText(text: String?) = apply {
            text?.takeIf { it.isNotEmpty() }?.let { components.add(TextDisplay.of(it)) }
        }

        fun addDivider(isLarge: Boolean = false, isInvisible: Boolean = false) = apply {
            val spacing = if (isLarge) Separator.Spacing.LARGE else Separator.Spacing.SMALL
            val separator = if (isInvisible) {
                Separator.createInvisible(spacing)
            } else {
                Separator.createDivider(spacing)
            }
            components.add(separator)
        }

        fun addDivider(isLarge: Boolean = false) = apply {
            val spacing = if (isLarge) Separator.Spacing.LARGE else Separator.Spacing.SMALL
            components.add(Separator.createDivider(spacing))
        }

        fun addSection(thumbnailUrl: String, text: String) = apply {
            components.add(Section.of(Thumbnail.fromUrl(thumbnailUrl), TextDisplay.of(text)))
        }

        fun addSection(button: Button, text: String) = apply {
            components.add(Section.of(button, TextDisplay.of(text)))
        }

        fun addRow(row: ActionRow) = apply {
            components.add(row)
        }

        fun addImage(url: String) = apply {
            components.add(MediaGallery.of(MediaGalleryItem.fromUrl(url)))
        }

        fun withColor(hexColor: String?) = apply {
            hexColor?.takeIf { it.isNotEmpty() }?.let {
                val colorStr = if (it.startsWith("#")) it else "#$it"
                this.accentColor = Color.decode(colorStr)
            }
        }

        fun setEphemeral(ephemeral: Boolean) = apply {
            this.isEphemeral = ephemeral
        }

        fun disableMentions() = apply {
            this.disallowedMentions = true
        }

        fun build(): Container {
            var container = Container.of(components)
            accentColor?.let { container = container.withAccentColor(it) }
            return container
        }

        // === Métodos de reply ===
        fun reply(event: SlashCommandInteractionEvent) = apply {
            if (event.isAcknowledged) {
                // Se já tem defer, edita a original via Hook
                val edit = event.hook.editOriginalComponents(build())
                    .useComponentsV2()
                // Nota: setEphemeral não funciona aqui pois já foi definido no defer

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            } else {
                // Se não tem defer, responde criando a mensagem
                val reply = event.replyComponents(build())
                    .useComponentsV2()
                    .setEphemeral(isEphemeral)

                if (disallowedMentions) {
                    reply.setAllowedMentions(emptyList()).queue()
                } else {
                    reply.queue()
                }
            }
        }

        fun reply(event: ButtonInteractionEvent) = apply {
            if (event.isAcknowledged) {
                val edit = event.hook.editOriginalComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            } else {
                val reply = event.replyComponents(build())
                    .useComponentsV2()
                    .setEphemeral(isEphemeral)

                if (disallowedMentions) {
                    reply.setAllowedMentions(emptyList()).queue()
                } else {
                    reply.queue()
                }
            }
        }

        fun reply(event: EntitySelectInteractionEvent) = apply {
            if (event.isAcknowledged) {
                val edit = event.hook.editOriginalComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            } else {
                val reply = event.replyComponents(build())
                    .useComponentsV2()
                    .setEphemeral(isEphemeral)

                if (disallowedMentions) {
                    reply.setAllowedMentions(emptyList()).queue()
                } else {
                    reply.queue()
                }
            }
        }

        // Para Modal, geralmente se responde com texto ou embed, mas se seu build() retorna componentes:
        fun reply(event: ModalInteractionEvent) = apply {
            if (event.isAcknowledged) {
                val edit = event.hook.editOriginalComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            } else {
                val reply = event.replyComponents(build())
                    .useComponentsV2()
                    .setEphemeral(isEphemeral)

                if (disallowedMentions) {
                    reply.setAllowedMentions(emptyList()).queue()
                } else {
                    reply.queue()
                }
            }
        }

        // Método direto para Hook (sempre é edição)
        fun reply(hook: InteractionHook) = apply {
            val edit = hook.editOriginalComponents(build())
                .useComponentsV2()

            if (disallowedMentions) {
                edit.setAllowedMentions(emptyList()).queue()
            } else {
                edit.queue()
            }
        }


// === Métodos de editOriginal (Atualização da mensagem onde o componente está) ===

        fun editOriginal(event: ButtonInteractionEvent) = apply {
            if (event.isAcknowledged) {
                // Se usou deferEdit(), usamos o hook
                val edit = event.hook.editOriginalComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            } else {
                // Se não, editamos a mensagem diretamente via evento
                val edit = event.editComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            }
        }

        fun editOriginal(event: ModalInteractionEvent) = apply {
            if (event.isAcknowledged) {
                val edit = event.hook.editOriginalComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            } else {
                val edit = event.editComponents(build())
                    .useComponentsV2()

                if (disallowedMentions) {
                    edit.setAllowedMentions(emptyList()).queue()
                } else {
                    edit.queue()
                }
            }
        }

        companion object {
            @JvmStatic
            fun create(): ContainerBuilder = ContainerBuilder()
        }
    }
}